### PR TITLE
Adding cached KVs

### PR DESCRIPTION
**Pull request by gkroiz**
This PR adds the option to cache KVs for inference. Porting changes from https://github.com/Liyang90/lit-llama/tree/cached_kvs.

main changes:
- added KV caching
- moved rope cache to the whole model from each attention layer
- added mask cache to use for attention masking with cached kvs.
- fix ops and tensors that would cause recompilation.

### CODE REVIEW

**Comment by gkroiz** [comment]
cc @Liyang90 @carmocca

**Review by carmocca — COMMENTED** [review]
This is looking great!

File: generate.py [hunk 1]
```diff
@@ -50,19 +61,22 @@ def generate(
...
         logits = logits[0, -1] / temperature
 
         # optionally crop the logits to only the top k options
         if top_k is not None:
             v, _ = torch.topk(logits, min(top_k, logits.size(-1)))
-            logits[logits < v[[-1]]] = -float("Inf")
+            logits = torch.where(logits < v[[-1]], -float("Inf"), logits)
 
         probs = torch.nn.functional.softmax(logits, dim=-1)
         idx_next = torch.multinomial(probs, num_samples=1)
```
Comment (carmocca) [inline generate.py:73]: index_copy requires that the dtype matches, so we need

```suggestion
        idx_next = torch.multinomial(probs, num_samples=1).to(dtype=idx.dtype)
```

**Comment by carmocca** [comment]
I'm getting a 30% speedup :racehorse:

File: generate.py [hunk 2]
```diff
@@ -50,19 +61,22 @@ def generate(
         idx_cond = idx_cond if t <= max_seq_length else idx_cond[-max_seq_length:]
 
         # forward
-        logits = model(idx_cond.view(1, -1))
```
[the remark this answers is not in the document]
Comment (gkroiz) [reply generate.py:53]: You are right that `idx_cond` is not used and that we should trim the `cache_kvs` to `max_seq_length`. I will add these changes accordingly.

File: generate.py [hunk 3]
```diff
@@ -42,6 +42,17 @@ def generate(
...
+            torch.zeros((1, model.config.n_head, T_new, model.config.n_embd // model.config.n_head), device=idx.device),
+        )
+        for _ in range(model.config.n_layer)
+    ]
+
+    input_pos = torch.arange(0, T).to(idx.device)
+    input_idx = idx.index_select(0, input_pos)
+
     # generate max_new_tokens tokens
     for t in range(T, T_new):
```
[the remark this answers is not in the document]
Comment (gkroiz) [reply generate.py:55]: When I try using what is suggested above, the generation runtime is much slower. To fix this, I changed `idx[start:stop]` to `idx.index_select(0, input_pos)` and redefine `input_pos`. However, once these changes are made, this loop is basically the same as the `for loop` that is currently used.

File: lit_parrot/model.py [hunk 4]
```diff
@@ -110,27 +170,30 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
         q = torch.cat((q_roped, q[..., n_elem:]), dim=-1)
         k = torch.cat((k_roped, k[..., n_elem:]), dim=-1)
 
+        if cache_kv is not None:
+            cache_k, cache_v = cache_kv
+            cache_k = cache_k.index_copy(2, input_pos, k)
+            cache_v = cache_v.index_copy(2, input_pos, v)
+            cache_kv = cache_k, cache_v
+            k = cache_k[:]
+            v = cache_v[:]
```
[the remark this answers is not in the document]
Comment (gkroiz) [reply lit_parrot/model.py]: I don't think we need to create a hard copy since `k` and `v` are used for `scaled_dot_product_attention` which shouldn't edits the inputs. I made the adjustment, we can revert if needed.

@Liyang90 do you think this is needed?

File: lit_parrot/model.py [hunk 5]
```diff
@@ -110,27 +170,41 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
+        cos, sin = rope
         q_roped = apply_rope(q[..., :n_elem], cos, sin)
         k_roped = apply_rope(k[..., :n_elem], cos, sin)
         q = torch.cat((q_roped, q[..., n_elem:]), dim=-1)
         k = torch.cat((k_roped, k[..., n_elem:]), dim=-1)
 
+        if kv_cache is not None:
+            cache_k, cache_v = kv_cache
+            # check if reached token limit
+            if input_pos[-1] > max_new_tokens:
```
Comment (gkroiz) [inline lit_parrot/model.py:183]: Added token limit logic here cc @carmocca

Comment (carmocca) [reply lit_parrot/model.py:183]: @gkroiz I think there was a mixup between `max_new_tokens` and `max_seq_length`.

- `max_new_tokens` defines how many tokens to generate
- `max_seq_length` defines what is the block size allowed for the model

My comment at https://github.com/Lightning-AI/lit-parrot/pull/51#discussion_r1190554667 referred to the latter.

Also, did you mean "increases overall inference time"?

Comment (gkroiz) [reply lit_parrot/model.py:183]: Yes, I meant increases overall inference time, I'll change the original comment

Comment (gkroiz) [reply lit_parrot/model.py]: fixed the variable discrepancy.

File: lit_parrot/model.py [hunk 6]
```diff
@@ -44,17 +49,50 @@ def _init_weights(self, module: nn.Module) -> None:
...
         x = self.transformer.wte(idx)  # token embeddings of shape (b, t, n_embd)
 
-        for block in self.transformer.h:
-            x = block(x)
+        if input_pos is None:  # proxy for use_cache=False
+            for block in self.transformer.h:
+                x, _ = block(x, (cos, sin), mask, max_seq_length)
+        else:
+            if not self.kv_caches:
+                cache_shape = (B, self.config.n_head, max_seq_length, self.config.n_embd // self.config.n_head)
```
Comment (Liyang90) [inline lit_parrot/model.py:88]: What if `B` or `max_seq_length` changes for later calls to `forward` for another batch of generation inputs?

Comment (gkroiz) [reply lit_parrot/model.py:88]: This is a good point, maybe we can reset the cache if we notice a difference in `B` or `max_seq_length`

File: generate.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (carmocca) [reply generate.py:56]: What about using `idx.narrow(0, start, stop - start)`. Is that also slower? (updated snippet above)

File: lit_parrot/model.py (hunk 6 above)
Comment (Liyang90) [reply lit_parrot/model.py:88]: That may work, but it would complicate the code, and may cause unintentional recompilation.

That's the initial concern that leading to me putting cache-kvs as inputs and outputs. It prevents input related parameters being put into the model.

Comment (carmocca) [reply lit_parrot/model.py:88]: I think it's fair to have the user do `model.kv_caches.clear()` if these parameter change

Comment (gkroiz) [reply lit_parrot/model.py:88]: On second thought, don't we need to reset the cache everytime we have a new batch of generation inputs? If so, it wouldn't matter whether `B` and `max_seq_length` match.

Comment (gkroiz) [reply lit_parrot/model.py:88]: > I think it's fair to have the user do `model.kv_caches.clear()` if these parameters change

This might be the best solution, where it is up to the user to clear the cache after each batch.

File: generate.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (gkroiz) [reply generate.py:60]: The updated snippet above is also much slower.

[the remark this answers is not in the document]
Comment (carmocca) [reply generate.py]: Okay. I gave it a shot and ended up with https://github.com/Lightning-AI/lit-parrot/pull/51/commits/e07127bdaa371a6ecff04f79ce129264717db6bd which does seem to match in performance. Resolving

File: generate.py
[the remark this answers is not in the document]
Comment (carmocca) [reply generate.py:1]: I made the changes to chat.py to use the KV cache.
I get a speedup running on CUDA and XLA.

However, on XLA, it seems like there's recompilation after every message as it takes about ~45s to reply even after the first one.

It would be very nice if we could avoid that. But that can be explored in a follow-up PR.

[the remark this answers is not in the document]
Comment (carmocca) [reply generate.py:1]: Yes

**Review by carmocca — APPROVED** [review]
We're done here! Awesome

File: lit_parrot/model.py [hunk 7]
```diff
@@ -14,6 +13,9 @@
 
 from lit_parrot.config import Config
 
+RopeCache = Tuple[torch.Tensor, torch.Tensor]
+KvCache = Tuple[torch.Tensor, torch.Tensor]
```
Comment (lantiga) [inline lit_parrot/model.py:17]: Super minor nit:

`RoPECache` -> `RoPECache`
`KvCache` -> `KVCache`

**Review by lantiga — APPROVED** [review]
Looks amazing! Good to merge (I just proposed a change in capitalization of two types, just if we feel like it)

File: lit_parrot/model.py [hunk 8]
```diff
@@ -44,17 +49,57 @@ def _init_weights(self, module: nn.Module) -> None:
...
+            self.rope_cache = self.build_rope_cache(idx)
+        if self.mask_cache is None:
+            self.mask_cache = self.build_mask_cache(idx)
+
+        cos, sin = self.rope_cache
+        if input_pos is not None:
+            cos = cos.index_select(0, input_pos)
+            sin = sin.index_select(0, input_pos)
+            mask = self.mask_cache.index_select(2, input_pos)
+            mask = mask[:, :, :max_seq_length, :max_seq_length]
```
Comment (Liyang90) [inline lit_parrot/model.py:74]: Shouldn't this be:
`mask = mask[:, :, :, :max_seq_length]`
following the changes in https://github.com/Lightning-AI/lit-llama/compare/main...Liyang90:lit-llama:cached_kvs

File: generate.py [hunk 9]
```diff
@@ -35,42 +36,49 @@ def generate(
...
+        idx_next = torch.multinomial(probs, num_samples=1).to(dtype=dtype)
 
+        # advance
+        input_pos = input_pos[-1:] + 1
         # concatenate the new generation
-        idx[t] = idx_next
+        idx = idx.index_copy(0, input_pos, idx_next)
 
         # if <eos> token is triggered, return the output (stop generation)
         if idx_next == eos_id:
```
Comment (Liyang90) [inline generate.py:80]: Checking the value of `idx_next` is better done after `xm.mark_step()`, because it is a lazy tensor.

File: chat.py [hunk 10]
```diff
@@ -26,53 +29,70 @@ def generate(
...
         import torch_xla.core.xla_model as xm
+
         xm.mark_step()
 
-    for t in range(T, max_seq_length):
+    yield_i = -1
+    for t in range(max_new_tokens):
         # forward
-        logits = model(idx.view(1, -1))
+        logits = model(idx.view(1, -1), max_seq_length, input_pos)
```
Comment (Liyang90) [inline chat.py:61]: I can see why there may still be recompilation here. If the ops below cannot be made static, adding a `xm.mark_step()` right after `logits = model(idx.view(1, -1), max_seq_length, input_pos)` may alleviate the recompilation a lot, because now only the small amount of computation below is recompiling.

File: lit_parrot/model.py (hunk 8 above)
Comment (gkroiz) [reply lit_parrot/model.py:74]: Thanks for the review. Following up on your question, the code currently looks like:
```
...
            mask = mask[:, :, :, :max_seq_length]
        else:
            cos = cos[:T]
            sin = sin[:T]
            mask = self.mask_cache[:, :, :T, :T]
...
```
Like you mentioned, the changes in https://github.com/Lightning-AI/lit-llama/compare/main...Liyang90:lit-llama:cached_kvs only trimmed the last dimension in `mask`. What is the logic behind trimming only the last dimension in `mask` when using `input_pos` and trimming the last two dimensions when `input_pos == None`?

File: chat.py (hunk 10 above)
Comment (carmocca) [reply chat.py:61]: I think you are right. Just gave it a try and seems faster.

But why after forwards instead of after `multinomial`? Isn't the issue mostly the dynamic nature of the yielding logic?

Also, is there a guide on `mark_step` recommendations and best practices?

Comment (Liyang90) [reply chat.py:61]: Yes. putting it after `idx = torch.multinomial(probs, num_samples=1)` probably also works. It needs to be before `buffer[min(t, buffer_length - 1)] = idx` at least.

There is no guide unfortunately, it's mainly from the nature of lazy tensor, and what `xm.mark_step()` do is capture all the lazy tensors at that moment and sync them through graph execution.

File: lit_parrot/model.py (hunk 8 above)
Comment (Liyang90) [reply lit_parrot/model.py:74]: because `mask = self.mask_cache.index_select(2, input_pos)` already selects the proper rows on dim 2.

File: generate.py (hunk 9 above)
Comment (gkroiz) [reply generate.py:84]: Fixed

File: lit_parrot/model.py (hunk 8 above)
Comment (gkroiz) [reply lit_parrot/model.py]: Fixed

File: chat.py (hunk 10 above)
Comment (gkroiz) [reply chat.py:60]: Moved the `xl.mark_step()` to before the buffer for consistency with `generate.py`. Thanks @Liyang90 for the explanation!
